### PR TITLE
Custom errors for ERC tokens

### PR DESCRIPTION
**Pull request by ernestognw**
## Description

Created EIP-draft for the errors to be thrown in the standard token implementations (ERC20, ERC721, ERC1155).

### CODE REVIEW

File: EIPS/eip-custom_errors_for_erc_tokens.md [hunk 1]
```diff
@@ -0,0 +1,532 @@
...
+category: ERC
+requires: 20, 721, 1155
+created: <date created on, in ISO 8601 (yyyy-mm-dd) format>
+---
+
+## Abstract
+
+The following specification allows for the use of a standard list of Solidity [custom errors](https://blog.soliditylang.org/2021/04/21/custom-errors/) to be used within [EIP-20](./eip-20.md), [EIP-721](./eip-721.md) and [EIP-1155](./eip-1155.md) tokens.
+
+Ethereum applications (DApps) and wallets have historically used [revert statements](https://docs.soliditylang.org/en/v0.8.16/control-structures.html#revert-statement) to show relevant failure data to the users, however, this EIP provides a standard list of errors designed to give at least the same relevant information, but in a structured and expected way.
```
Comment (frangio) [inline EIPS/eip-custom_errors_for_erc_tokens.md:18]:
```suggestion
Ethereum applications (DApps) and wallets have historically relied on [revert reason strings](https://docs.soliditylang.org/en/v0.8.16/control-structures.html#revert-statement) to show the cause of transaction errors to users. More recent Solidity versions offer rich revert reasons that require error-specific decoding. This EIP defines a standard set of errors designed to give at least the same relevant information as revert reason strings, but in a structured and expected way that clients can implement decoding for.
```

File: EIPS/eip-custom_errors_for_erc_tokens.md [hunk 2]
```diff
@@ -0,0 +1,532 @@
...
+
+## Motivation
+
+Since the introduction of Solidity custom errors in [v0.8.4](https://github.com/ethereum/solidity/releases/tag/v0.8.4), these have provided a way to show failures in a more expresive way with dynamic arguments, while reducing deployment costs.
+
+At the moment of the release of custom errors, the standard tokens ([EIP-20](./eip-20.md), [EIP-721](./eip-721.md), [EIP-1155](./eip-1155.md)) were already in finalized state, so no error specification was included.
+
+An error specification will allow users to expect more consistent error messages across applications or testing environments, while exposing pertinent arguments and overall reducing the need of writing expensive revert strings in the deployment bytecode.
+
+The list of token standard errors is defined in the specification below.
```
Comment (frangio) [inline EIPS/eip-custom_errors_for_erc_tokens.md:28]: This is redundant.

```suggestion
```

File: EIPS/eip-custom_errors_for_erc_tokens.md [hunk 3]
```diff
@@ -0,0 +1,532 @@
...
+
+The following specification allows for the use of a standard list of Solidity [custom errors](https://blog.soliditylang.org/2021/04/21/custom-errors/) to be used within [EIP-20](./eip-20.md), [EIP-721](./eip-721.md) and [EIP-1155](./eip-1155.md) tokens.
+
+Ethereum applications (DApps) and wallets have historically used [revert statements](https://docs.soliditylang.org/en/v0.8.16/control-structures.html#revert-statement) to show relevant failure data to the users, however, this EIP provides a standard list of errors designed to give at least the same relevant information, but in a structured and expected way.
+
+## Motivation
+
+Since the introduction of Solidity custom errors in [v0.8.4](https://github.com/ethereum/solidity/releases/tag/v0.8.4), these have provided a way to show failures in a more expresive way with dynamic arguments, while reducing deployment costs.
+
+At the moment of the release of custom errors, the standard tokens ([EIP-20](./eip-20.md), [EIP-721](./eip-721.md), [EIP-1155](./eip-1155.md)) were already in finalized state, so no error specification was included.
```
Comment (frangio) [inline EIPS/eip-custom_errors_for_erc_tokens.md:24]:
```suggestion
At the moment of the release of custom errors, the standard tokens ([EIP-20](./eip-20.md), [EIP-721](./eip-721.md), [EIP-1155](./eip-1155.md)) were already in finalized state, so no errors are included in their specifications.
```

File: EIPS/eip-custom_errors_for_erc_tokens.md [hunk 4]
```diff
@@ -0,0 +1,504 @@
...
+
+The provided interfaces with errors SHOULD conform with the [error grammar rules](#error-grammar-rules) described in the [rationale](#rationale) section.
+
+It's important to note that errors MAY be thrown in each scenario, but that's up to the implementers, unless a revert reason is specified in their corresponding EIPs.
+
+This EIP aims to provide a clear standard only in case an error will be thrown.
+
+### [EIP-20](./eip-20.md)
+
+#### `ERC20InvalidSender(address)`
```
Comment (frangio) [inline EIPS/eip-custom_errors_for_erc_tokens.md:42]: We should include parameter names here:

```suggestion
#### `ERC20InvalidSender(address sender)`
```

File: EIPS/eip-custom_errors_for_erc_tokens.md [hunk 5]
```diff
@@ -0,0 +1,504 @@
...
+
+Indicates a failure with the token sender.
+
+##### Parameters
+
+- `address _sender`: The address of whose tokens are transferred from.
+
+Throw when the cause of failure is the sender in a transfer.
```
Comment (frangio) [inline EIPS/eip-custom_errors_for_erc_tokens.md:50]: The term "throw" is taken from other languages, it doesn't really show up in Solidity, so I would avoid it. Here we could say "Used when ..."

File: EIPS/eip-custom_errors_for_erc_tokens.md [hunk 6]
```diff
@@ -0,0 +1,504 @@
...
+
+- MUST NOT be thrown for approval operations.
+- MUST NOT be thrown for balance or allowance requirements.
+  - Use `ERC20InsufficientBalance` or `ERC20InsufficientAllowance` instead.
+- MUST be thrown for unintended transfers from the zero address.
```
Comment (frangio) [inline EIPS/eip-custom_errors_for_erc_tokens.md:55]: Let's list the "musts" first, then the "must nots".

Here "unintended" sounds weird. Maybe "disallowed"?

File: EIPS/eip-custom_errors_for_erc_tokens.md [hunk 7]
```diff
@@ -0,0 +1,504 @@
...
+- `address _sender`: The address of whose tokens are transferred from.
+- `uint256 _balance`: The current amount of token the sender has.
+- `uint256 _needed`: The amount of token the sender needs to perform the operation.
+
+Throw when the cause of failure is the balance of a sender.
+
+- MUST NOT be thrown if `_balance` is equal or greater than `_needed`.
+- MUST be thrown when `_balance` is less than `_needed`.
+
+#### `ERC20InvalidApprover(address)`
```
Comment (frangio) [inline EIPS/eip-custom_errors_for_erc_tokens.md:86]: I think we need to leave this one out of the spec, even though we might use it in OpenZeppelin Contracts. It would never show up in a standard ERC20 token because the approver/owner is always `msg.sender`.

File: EIPS/eip-custom_errors_for_erc_tokens.md [hunk 8]
```diff
@@ -0,0 +1,504 @@
...
+
+### [EIP-20](./eip-20.md)
+
+#### `ERC20InvalidSender(address)`
+
+Indicates a failure with the token sender.
+
+##### Parameters
+
+- `address _sender`: The address of whose tokens are transferred from.
```
Comment (frangio) [inline EIPS/eip-custom_errors_for_erc_tokens.md:48]: IMO parameters are mostly self-explanatory and I would remove this section. If we need to explain their meaning, I would add a small section at the beginning of the EIP-20 section saying "here is what we mean when we say sender, receiver, spender, etc", and we can also say what we mean by the "needed" parameter in general.

File: EIPS/eip-custom_errors_for_erc_tokens.md [hunk 9]
```diff
@@ -0,0 +1,504 @@
...
+
+- `address _receiver` The address of whose tokens are transferred to.
+
+Throw when the cause of failure is the receiver in a transfer.
+
+- MUST NOT be thrown for approval operations.
+- MUST be thrown for unintended transfers to the zero address.
+- MUST be thrown for unintended transfers to non-compatible addresses (eg. contract addresses).
+
+#### `ERC20InsufficientBalance(address, uint256, uint256)`
```
Comment (frangio) [inline EIPS/eip-custom_errors_for_erc_tokens.md:71]: I would list errors in order to most common to least common. So insufficient balance should be the first one, then insufficient allowance, then the others.

File: EIPS/eip-custom_errors_for_erc_tokens.md [hunk 10]
```diff
@@ -0,0 +1,488 @@
...
+status: Draft
+type: Standards Track
+category: ERC
+requires: 20, 721, 1155
+created: <date created on, in ISO 8601 (yyyy-mm-dd) format>
+---
+
+## Abstract
+
+The following specification allows for the use of a standard list of Solidity [custom errors](https://blog.soliditylang.org/2021/04/21/custom-errors/) to be used within [EIP-20](./eip-20.md), [EIP-721](./eip-721.md) and [EIP-1155](./eip-1155.md) tokens.
```
Comment (frangio) [inline EIPS/eip-custom_errors_for_erc_tokens.md:16]: EIP editors will ask us to remove all external links. https://eips.ethereum.org/EIPS/eip-1#linking-to-external-resources

Generally the policy is:

>  If it is possible to implement a standard without a linked resource, then the linked resource is unnecessary.

File: EIPS/eip-custom_errors_for_erc_tokens.md (hunk 4 above)
Comment (ernestognw) [reply EIPS/eip-custom_errors_for_erc_tokens.md:41]: I remove them because some of them became very large, (eg. `ERC1155InsufficientBalance`).
Not sure if it's the best reading experience, also the parameters are outlined below each error.

What do you think?

File: EIPS/eip-custom_errors_for_erc_tokens.md (hunk 5 above)
Comment (ernestognw) [reply EIPS/eip-custom_errors_for_erc_tokens.md]: Agree overall, except with the alternative "Used".

I changed to "Used" because I have no better alternative but I'd think in something more specific to errors.

File: EIPS/eip-custom_errors_for_erc_tokens.md (hunk 7 above)
Comment (ernestognw) [reply EIPS/eip-custom_errors_for_erc_tokens.md:86]: I was about to say the same in your edit.
We don't know the specific use case of an ERC20. Maybe an approver might be blacklisted and we can't know it or enforce otherwise. I agree we're providing a clear go-to when those cases happen.

File: EIPS/eip-custom_errors_for_erc_tokens.md [hunk 11]
```diff
@@ -0,0 +1,488 @@
...
+
+## Specification
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119 and RFC 8174.
+
+The provided interfaces with errors SHOULD conform with the [error grammar rules](#error-grammar-rules) described in the [rationale](#rationale) section.
+
+It's important to note that errors MAY be thrown in each scenario, but that's up to the implementers, unless a revert reason is specified in their corresponding EIPs.
+
+This EIP aims to provide a clear standard only in case an error will be thrown.
```
Comment (frangio) [inline EIPS/eip-custom_errors_for_erc_tokens.md:38]: This part (especially the first paragraph) isn't clear. I think what you mean to say is something like this:

```suggestion
This EIP defines standard errors that may be used by implementations in certain scenarios, but does not specify whether implementations should revert in those scenarios, which remains up to the implementers, unless a revert is mandated by the corresponding EIPs.
```

File: EIPS/eip-custom_errors_for_erc_tokens.md (hunk 9 above)
Comment (ernestognw) [reply EIPS/eip-custom_errors_for_erc_tokens.md:71]: I'd define clearly what `most common` means.

Overall, the errors are following this arbitrary order:

First, transfer related:
1. Sender (address)
2. Receiver (address)
3. Balance (number)

Then, approval related:
4. Approver (address)
5. Spender (address)
6. Approval (number)

If we're moving `Balance` to the very beginning, I'd move `Approval` to the number 4 for consistency.
And, if we break the consistency, I'd put a formal definition of what `common` means.

What do you think?

File: EIPS/eip-custom_errors_for_erc_tokens.md (hunk 10 above)
Comment (ernestognw) [reply EIPS/eip-custom_errors_for_erc_tokens.md]: I removed all of the external links.
Do any of them sound required to you? So I can recover it.

File: EIPS/eip-custom_errors_for_erc_tokens.md (hunk 11 above)
Comment (ernestognw) [reply EIPS/eip-custom_errors_for_erc_tokens.md]: Yep, you got it right. I'm adding your suggestion :)

File: EIPS/eip-custom_errors_for_erc_tokens.md (hunk 8 above)
Comment (ernestognw) [reply EIPS/eip-custom_errors_for_erc_tokens.md]: > IMO parameters are mostly self-explanatory
Agree.

But there's something that's feeling off for me here.

Adding a section where we define subjects makes me feel that we should define the list of subjects for this EIP, which is something we previously stated not to do. Still, `_needed` is not a subject so I'm not sure my argument is completely correct.

I'll think about it but I'm sharing in case you have something to add :)

File: EIPS/eip-custom_errors_for_erc_tokens.md (hunk 4 above)
Comment (frangio) [reply EIPS/eip-custom_errors_for_erc_tokens.md:40]: I really think parameters names are necessary in the title. Haven't seen how large it becomes for insufficient balance but if it is really that large I'd consider giving that function special treatmen.

File: EIPS/eip-custom_errors_for_erc_tokens.md (hunk 8 above)
Comment (frangio) [reply EIPS/eip-custom_errors_for_erc_tokens.md]: >something we previously stated not to do

I didn't see them as belonging in the rationale section, but I think it would make sense in the specification section as documentation.

File: EIPS/eip-custom_errors_for_erc_tokens.md (hunk 4 above)
Comment (ernestognw) [reply EIPS/eip-custom_errors_for_erc_tokens.md]: No problem, I added it. The title is large but I don't think that's a problem.

File: EIPS/eip-custom_errors_for_erc_tokens.md (hunk 9 above)
Comment (ernestognw) [reply EIPS/eip-custom_errors_for_erc_tokens.md]: I'm going for the second one

File: EIPS/eip-custom_errors_for_erc_tokens.md (hunk 8 above)
Comment (ernestognw) [reply EIPS/eip-custom_errors_for_erc_tokens.md]: I just created a glossary table. I'd think it's expressive enough but something feels weird. I'd appreciate help with the wording.

File: EIPS/eip-custom_errors_for_erc_tokens.md [hunk 12]
```diff
@@ -0,0 +1,401 @@
...
+Used in approvals.
+
+- MUST be used for disallowed approvals to the zero address.
+- MUST be used for disallowed approvals to the owner itself.
+- MUST NOT be used for transfer operations.
+  - Use `ERC20InsufficientAllowance` instead.
+
+### [EIP-721](./eip-721.md)
+
+#### `ERC721InvalidOwner(address sender, uint256 tokenId)`
```
Comment (frangio) [inline EIPS/eip-custom_errors_for_erc_tokens.md:95]:
```suggestion
#### `ERC721InvalidOwner(address sender, address needed, uint256 tokenId)`
```

Comment (ernestognw) [reply EIPS/eip-custom_errors_for_erc_tokens.md]: Applied second suggestion with name `owner` on the last commit

File: EIPS/eip-custom_errors_for_erc_tokens.md [hunk 13]
```diff
@@ -0,0 +1,393 @@
...
+
+- **Transfer**: An operation in which a _sender_ moves any form of _balance_ to a _receiver_.
+- **Approval**: An operation in which an _approver_ grants any form of _approval_ to an _operator_.
+
+Whose subjects outlined are expected to be exhaustive to represent _what_ can go wrong in a transaction, which can be shown as an error by adding a [failure prefix](#error-prefixes).
+
+Note that `Token` or `TokenID` MUST NOT be a subject for the following reasons:
+
+1. The [Domain](#domain) is explicit enough to identify it as a token error.
+2. Although `TokenID` may identify an error, the token itself SHOULD NOT the root of the error, but can be additional information.
```
Comment (frangio) [inline EIPS/eip-custom_errors_for_erc_tokens.md:250]: I don't buy point 2 so strongly. I can imagine an ERC721 NFT with a supply of 10k tokens where transfers of token id > 10k result in an "InvalidTokenId" error. In fact, in this case what would be the error we'd use in OpenZeppelin's implementation? "InvalidOwner" isn't accurate because there is no "current owner" for an unminted token, address(0) is not considered a valid account so I wouldn't say it's the current owner. Maybe we need "InvalidTokenId" for this case!

If so, I would rephrase this part to say "The `Token` is never considered the subject of an error, but it is rather the context in which the error happens, as identified by the Domain."

File: EIPS/eip-custom_errors_for_erc_tokens.md [hunk 14]
```diff
@@ -0,0 +1,393 @@
...
+### Error prefixes
+
+An error prefix is what complements a subject, so it can be read as an error.
+Developers can think about an error prefix as the _why_ an error happened.
+
+A prefix can be `Invalid` for general incorrectness, or more specific like `Insufficient` for amounts (or lack of).
+
+### Domain
+
+Although these errors are considered to be expressive enough, each error's arguments may vary depending on the token domain, causing a `DeclarationError` if there are errors with the same name and different arguments.
```
Comment (frangio) [inline EIPS/eip-custom_errors_for_erc_tokens.md:272]:
```suggestion
Each error's arguments may vary depending on the token domain. If there are errors with the same name and different arguments, the Solidity compiler currently fails with a `DeclarationError`.
```

File: EIPS/eip-custom_errors_for_erc_tokens.md [hunk 15]
```diff
@@ -0,0 +1,393 @@
...
+For that reason, a domain prefix is proposed to avoid declaration clashing, which is the name of the ERC and its corresponding number appended at the beginning.
+
+Example:
+
+```solidity
+ERC20InsufficientApproval(address spender, uint256 allowance, uint256 needed);
+ERC721InsufficientApproval(address operator, uint256 tokenId);
+```
+
+A prefix is also RECOMMENDED for cases when there's
```
Comment (frangio) [inline EIPS/eip-custom_errors_for_erc_tokens.md:290]: The sentence is incomplete

File: EIPS/eip-custom_errors_for_erc_tokens.md [hunk 16]
```diff
@@ -0,0 +1,393 @@
...
+
+### Arguments
+
+The selection of arguments depends on the subject involved, and it SHOULD follow the order presented below:
+
+1. _Who_ is involved with the error (an `address sender`)
```
Comment (frangio) [inline EIPS/eip-custom_errors_for_erc_tokens.md:296]:
```suggestion
1. _Who_ is involved with the error (eg. `address sender`)
```

File: EIPS/eip-custom_errors_for_erc_tokens.md [hunk 17]
```diff
@@ -0,0 +1,393 @@
...
+2. _What_ failed (eg. `uint256 allowance`)
+3. _Why_ it failed, expressed in additional arguments (eg. `uint256 needed`)
+
+Consider arguments can overlap (eg. _Who_ is also _What_), so not all of the arguments will be present but the order SHOULD NOT be broken.
```
Comment (frangio) [inline EIPS/eip-custom_errors_for_erc_tokens.md:300]:
```suggestion
A particular argument may fall in overlapping categories (eg. _Who_ may also be _What_), so not all of these will be present but the order SHOULD NOT be broken.
```

File: EIPS/eip-custom_errors_for_erc_tokens.md [hunk 18]
```diff
@@ -0,0 +1,393 @@
...
+
+Note: Some tokens may need a `tokenId` or `id`. This is suggested to include at the end as additional information, according to the [subjects notes](#actions-and-subjects)
```
Comment (frangio) [inline EIPS/eip-custom_errors_for_erc_tokens.md:302]: Confused by this. Why don't we consider the token id part of the "what"?

File: EIPS/eip-custom_errors_for_erc_tokens.md [hunk 19]
```diff
@@ -0,0 +1,393 @@
...
+This EIP can not be enforced on non-upgradeable already deployed tokens, however, these tokens generally use similar conventions with small variations such as:
+
+- including/removing the [domain](#domain).
+- using different [error prefixes](#error-prefixes).
+- including similar [subjects](#actions-and-subjects).
+- changing the grammar order.
+
+Upgradeable contracts MAY be upgraded to implement this EIP.
+
+Implementers and DApp developers SHOULD expect these EIP errors to be present on new deployed contracts, but SHOULD handle variation errors for existing contracts, and revert strings that may be returned.
```
Comment (frangio) [inline EIPS/eip-custom_errors_for_erc_tokens.md:332]: This is my take:

```suggestion
Implementers and DApp developers that implement special support for tokens that are compliant with this EIP, SHOULD tolerate different errors emitted by non-compliant contracts, as well as classic revert strings.
```

File: EIPS/eip-custom_errors_for_erc_tokens.md [hunk 20]
```diff
@@ -0,0 +1,393 @@
...
+
+At the moment of the release of custom errors, the standard tokens ([EIP-20](./eip-20.md), [EIP-721](./eip-721.md), [EIP-1155](./eip-1155.md)) were already in finalized state, so no errors are included in their specification.
+
+An error specification will allow users to expect more consistent error messages across applications or testing environments, while exposing pertinent arguments and overall reducing the need of writing expensive revert strings in the deployment bytecode.
+
+## Specification
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119 and RFC 8174.
+
+The provided interfaces with errors SHOULD conform with the [error grammar rules](#error-grammar-rules) described in the [rationale](#rationale) section.
```
[the remark this answers is not in the document]
Comment (ernestognw) [reply EIPS/eip-custom_errors_for_erc_tokens.md:32]: I'll accept this but I want to chat about what I'm intending to do here with the `SHOULD`s, but I think my approach was not the best.

Resolving

File: EIPS/eip-custom_errors_for_erc_tokens.md [hunk 21]
```diff
@@ -0,0 +1,393 @@
...
+
+| Name        | Description                                                                         |
+| ----------- | ----------------------------------------------------------------------------------- |
+| `sender`    | The address of whose token(s) (is/are) transferred from.                            |
+| `balance`   | Current balance for the interacting account.                                        |
+| `needed`    | Amount minimum required to perform an action.                                       |
+| `receiver`  | Address owner of the token(s) transferred.                                          |
+| `spender`   | Address not owner of the token(s) transferred, but allowed to operate on (it/them). |
+| `allowance` | Amount of token(s) a `spender` is allowed to operate with.                          |
+| `approver`  | Owner of the token(s) being approved to an `spender`.                               |
```
[the remark this answers is not in the document]
Comment (ernestognw) [reply EIPS/eip-custom_errors_for_erc_tokens.md:219]: I don't like the idea of `approve` as if it was referencing the function, which is not there for an ERC1155, for example.

I'd keep it but with "approval" instead of `approve`

```suggestion
| `approver`  | Address initiating an approval operation.                               |
```

File: EIPS/eip-custom_errors_for_erc_tokens.md [hunk 22]
```diff
@@ -0,0 +1,393 @@
...
+
+Any addition to this EIP is considerred a new proposal, but SHOULD follow the guidelines presented in the [rationale](#rationale) section to keep consistency.
+
+Implementation specific errors (such as extensions) SHOULD also follow the same guidelines, although subjects and arguments may vary.
+
+## Rationale
+
+The objectives of creating a standard for token errors are to provide context about the error, while keeping gramatical sense, and making a moderated use of meaningful arguments.
+
+Considering the outlined above, the errors are designed based on the standard actions that can be performed on each token, and the [subjects](#actions-and-subjects) involved.
```
[the remark this answers is not in the document]
Comment (ernestognw) [reply EIPS/eip-custom_errors_for_erc_tokens.md:236]: I'm accepting since I consider your suggestion way better than my original content. But, I'd like to discuss the objectives with you.

My intention with 'keeping grammatical sense" was basically to make the errors _readable_, and not end up with something like `TransferFailedNotOwner(...)`.

The proposed grammar is a direct consequence of this objective. We can keep it as your suggestion but I'd like to consider the readability part of the objectives.

File: EIPS/eip-custom_errors_for_erc_tokens.md [hunk 23]
```diff
@@ -0,0 +1,393 @@
...
+| ----------- | ----------------------------------------------------------------------------------- |
+| `sender`    | The address of whose token(s) (is/are) transferred from.                            |
+| `balance`   | Current balance for the interacting account.                                        |
+| `needed`    | Amount minimum required to perform an action.                                       |
+| `receiver`  | Address owner of the token(s) transferred.                                          |
+| `spender`   | Address not owner of the token(s) transferred, but allowed to operate on (it/them). |
+| `allowance` | Amount of token(s) a `spender` is allowed to operate with.                          |
+| `approver`  | Owner of the token(s) being approved to an `spender`.                               |
+| `tokenId`   | The identifier number of a token type.                                              |
+| `owner`     | Address of the owner of a token type.                                               |
```
[the remark this answers is not in the document]
Comment (ernestognw) [reply EIPS/eip-custom_errors_for_erc_tokens.md]: I think `token` vs `tokenType` is the same case as with `tokens` vs `token(s)`

The former is already considering all variants, either one token or many since it's plural and we kept `tokens` in other paragraphs.

I'd root for keeping `toke type`  or `token` instead of `token or token type` since they represent the same.
I'm particularly biased towards `token type`.

File: EIPS/eip-custom_errors_for_erc_tokens.md [hunk 24]
```diff
@@ -0,0 +1,393 @@
...
+| `operator`  | Same as `spender`.                                                                  |
+| `id`        | Same as `tokenId`.                                                                  |
```
[the remark this answers is not in the document]
Comment (ernestognw) [reply EIPS/eip-custom_errors_for_erc_tokens.md]: I agree I like it better if we consolidate, but is a similar case of using `Approval` vs `Allowance`. We're just keeping consistency with the standard-defined name.

Although I'd agree with consolidating, I'd like to ask why is this consolidatable and not `Approval`? Friendly question 😛

File: EIPS/eip-custom_errors_for_erc_tokens.md [hunk 25]
```diff
@@ -0,0 +1,393 @@
...
+| `*Length`   | Array length for the prefixed parameter.                                            |
+
+### Error additions
+
+Any addition to this EIP is considerred a new proposal, but SHOULD follow the guidelines presented in the [rationale](#rationale) section to keep consistency.
+
+Implementation specific errors (such as extensions) SHOULD also follow the same guidelines, although subjects and arguments may vary.
```
[the remark this answers is not in the document]
Comment (ernestognw) [reply EIPS/eip-custom_errors_for_erc_tokens.md:230]: The second paragraph was added after our meeting yesterday. So far I'm trying to express the following:

1. I'm considering additions as new EIPs extending to this standard list, I'd like to express that for new EIPs, consistency is important
2. With implementation-specific errors, there's no enforcement of the grammar, **unless**, it's going to be proposed as a new EIP

Second paragraph can be deleted but I have the feeling I'm missing the idea. What do you think?

File: EIPS/eip-custom_errors_for_erc_tokens.md [hunk 26]
```diff
@@ -0,0 +1,393 @@
...
+
+## Specification
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119 and RFC 8174.
+
+The provided interfaces with errors SHOULD conform with the [error grammar rules](#error-grammar-rules) described in the [rationale](#rationale) section.
+
+This EIP defines standard errors that may be used by implementations in certain scenarios, but does not specify whether implementations should revert in those scenarios, which remains up to the implementers, unless a revert is mandated by the corresponding EIPs.
+
+The names of the error arguments follow the [parameter glossary](#parameter-glossary) section, defining a clear guideline for assigning values to them.
```
[the remark this answers is not in the document]
Comment (ernestognw) [reply EIPS/eip-custom_errors_for_erc_tokens.md:36]: Accepting with a small change:

```suggestion
The names of the error arguments are defined in the [Parameter Glossary](#parameter-glossary), and MUST be used according to those definitions.
```

File: EIPS/eip-custom_errors_for_erc_tokens.md [hunk 27]
```diff
@@ -0,0 +1,393 @@
...
+Considering the outlined above, the errors are designed based on the standard actions that can be performed on each token, and the [subjects](#actions-and-subjects) involved.
+
+### Actions and subjects
+
+The main actions that can be performed within a token are:
+
+- **Transfer**: An operation in which a _sender_ moves any form of _balance_ to a _receiver_.
+- **Approval**: An operation in which an _approver_ grants any form of _approval_ to an _operator_.
+
+Whose subjects outlined are expected to be exhaustive to represent _what_ can go wrong in a transaction, which can be shown as an error by adding a [failure prefix](#error-prefixes).
```
[the remark this answers is not in the document]
Comment (ernestognw) [reply EIPS/eip-custom_errors_for_erc_tokens.md]: Thank you so much!

File: EIPS/eip-custom_errors_for_erc_tokens.md (hunk 15 above)
Comment (ernestognw) [reply EIPS/eip-custom_errors_for_erc_tokens.md:290]: Thanks, missed that one. Removed

File: EIPS/eip-custom_errors_for_erc_tokens.md (hunk 16 above)
Comment (ernestognw) [reply EIPS/eip-custom_errors_for_erc_tokens.md:296]: I'd accept only if we agree an `address` is the only thing that can be there.

The rule is broken with `ERC1155InvalidArrayLength(uint256, uint256)`.

Comment (frangio) [reply EIPS/eip-custom_errors_for_erc_tokens.md:296]: What do you mean? "Eg." does not mean that what follows is the only option.

Re: array length, let's discuss that in the context of "who vs what". I think it's very clear in that example that the lengths are "what" and not "who"!

File: EIPS/eip-custom_errors_for_erc_tokens.md (hunk 23 above)
[the remark this answers is not in the document]
Comment (ernestognw) [reply EIPS/eip-custom_errors_for_erc_tokens.md]: Went for `token` only after sync meeting.

File: EIPS/eip-custom_errors_for_erc_tokens.md (hunk 24 above)
[the remark this answers is not in the document]
Comment (ernestognw) [reply EIPS/eip-custom_errors_for_erc_tokens.md]: Decided to consolidate for `tokenId` since the ERC1155 spec also mentions it, and parameter names are not considered as important as error names (for `Approval` vs `Allowance`).

File: EIPS/eip-custom_errors_for_erc_tokens.md (hunk 13 above)
Comment (ernestognw) [reply EIPS/eip-custom_errors_for_erc_tokens.md]: Section removed in favor of comments during our previous meeting.
Solving

File: EIPS/eip-custom_errors_for_erc_tokens.md [hunk 28]
```diff
@@ -0,0 +1,393 @@
...
+Consider that, in practice, some of the subjects in a contract might be called different while they represent the same. These are listed below:
+
+- **Balance**: Represents the access to a certain amount of tokens for an address.
+  - EIP-721: Instead of _balance_, an address is _owner_ of a token.
+- **Approval**: Represents a permission granted to an operator.
+  - EIP-20: Instead of _approval_, an operator has _allowance_.
+- **Operator**: Represents an address that has received approval over third-party tokens.
+  - EIP-20: Instead of _operator_, an address with allowance is an _spender_.
+
+If a subject is called different on a particular token standard, the error MUST be consistent with the standard's naming convention.
```
[the remark this answers is not in the document]
Comment (ernestognw) [reply EIPS/eip-custom_errors_for_erc_tokens.md]: Section removed in favor of additions from our previous meeting.

Solving.

File: EIPS/eip-custom_errors_for_erc_tokens.md (hunk 25 above)
[the remark this answers is not in the document]
Comment (ernestognw) [reply EIPS/eip-custom_errors_for_erc_tokens.md]: Merge both paragraphs.

File: EIPS/eip-custom_errors_for_erc_tokens.md [hunk 29]
```diff
@@ -0,0 +1,376 @@
...
+Used in approvals.
+
+- MUST be used for disallowed approvals to the zero address.
+- MUST be used for disallowed approvals to the owner itself.
+- MUST NOT be used for transfer operations.
+  - Use `ERC1155InsufficientApproval` instead.
+
+#### `ERC1155InvalidArrayLength(uint256 idsLength, uint256 valuesLength)`
+
+Indicates an array length mismatch between `_ids` and `_values` in a `safeBatchTransferFrom` operation.
```
Comment (ernestognw) [inline EIPS/eip-custom_errors_for_erc_tokens.md:204]:
```suggestion
Indicates an array length mismatch between `ids` and `values` in a `safeBatchTransferFrom` operation.
```

**Comment by ernestognw** [comment]
Merging to send PR to Ethereum's repo
